File: magicmachine.py
import pdb
import os
import sys
from itertools import izip_longest
from collections import deque
from datetime import datetime

class MagicMachine(object):
	class Halt(Exception): pass

	OPCODES = ('halt', 'set', 'push', 'pop', 'eq', 'gt', 'jmp', 'jt', 'jf', 
		'add', 'mult', 'mod', 'and', 'or', 'not', 'rmem', 'wmem', 'call', 
		'ret', 'out', 'in', 'noop')

	ADMIN_COMMANDS = ('memdump', 'quit', 'registers', 'set_register', 
		'record', 'stop_recording', 'fix teleporter')

	def __init__(self):
		self.memory = [0 for _ in xrange(2**15)]
		self.registers = [0 for _ in xrange(8)]
		self.stack = deque()
		self.input_buffer = None
		self.autoinput = deque()
		self.recording = False

		self.operations = []
		for opcode in self.OPCODES:
			op = getattr(self, "op_%s" % opcode)
			self.operations.append((op, op.__code__.co_argcount - 1))

	def load_bin(self, filename):
		with open(filename, "rb") as f:
			data = f.read()

		itr = iter(data)
		for address, (bit1, bit2) in enumerate(izip_longest(itr, itr)):
			self.memory[address] = ord(bit1) + (ord(bit2) * (2**8))

	def load_autoinput(self, filename):
		with open(filename, "r") as f:
			for line in f:
				self.autoinput.append(line.strip())

	def admin_record(self):
		print "recording..."
		self.recording = True

		with open("record.txt", "w") as f:
			f.write("started recording at %s\n" % \
				datetime.now().strftime("%Y-%m-%d %H:%M:%S"))

	def admin_stop_recording(self):
		print "stoped recording"
		self.recording = False

	def admin_registers(self):
		print self.registers

	def admin_set_register(self, register, value):
		print "setting register %s to %s" % (register, value)
		self.registers[int(register)] = int(value)

	def admin_fix_teleporter(self):
		self.registers[7] = 5
		self.memory[6027:6030] = [1, 32769, 32775] # r1 = r7
		self.memory[6030:6034] = [9, 32768, 32769, 1] # r0 = r1 + 1

	def admin_memdump(self, filename = None):
		n = 0

		if filename is None:
			while True:
				filename = "memdumps/memdump_%04d.txt" % n
				if os.path.exists(filename):
					n += 1
				else:
					break
		else:
			filename = "memdumps/%s.txt" % filename

		with open(filename, 'w') as f:
			f.write("Registers:\n\n")
			for n in xrange(8):
				f.write("%d: %d\n" % (n, self.registers[n]))

			f.write("\nStack:\n\n")
			if len(self.stack) == 0:
				f.write("(empty)")
			else:
				for s in self.stack:
					f.write("%s\n" % s)

			f.write("\n\nMemory:\n\n")
			for n in xrange(2**15):
				f.write("%05d: %d\n" % (n, self.memory[n]))

	def admin_quit(self):
		raise self.Halt()

	def run(self, start_address = 0):
		self.address = start_address

		while self.address < (2 ** 15):
			opaddr = self.address
			opnum = self.memory[self.address]
			operation, argcount = self.operations[opnum]

			opargs = []
			for _ in xrange(argcount):
				self.address += 1
				arg = self.memory[self.address]
				opargs.append(arg)

			if self.recording:
				with open("record.txt", "a") as f:
					f.write("%05d " % opaddr)
					f.write(self.OPCODES[opnum])

					for arg in opargs:
						f.write(' ')

						if arg >= (2**15):
							f.write("reg%d" % (arg%2**15))
						else:
							f.write("%d" % arg)

					f.write('\n')

			try:
				result = operation(*opargs)

				if result != 'jmp':
					self.address += 1

			except self.Halt:
				break

	def lookup(self, value):
		if value >= (2**15 + 8) or value < 0:
			raise ValueError(value)
		elif value >= (2**15):
			return self.registers[self.to_register(value)]
		else:
			return value

	def to_register(self, value):
		return value % (2**15)

	def op_halt(self):
		raise self.Halt()

	def op_set(self, a, b):
		self.registers[self.to_register(a)] = self.lookup(b)

	def op_push(self, a):
		self.stack.append(self.lookup(a))

	def op_pop(self, a):
		self.op_set(a, self.stack.pop())

	def op_eq(self, a, b, c):
		val = 1 if self.lookup(b) == self.lookup(c) else 0
		self.op_set(a, val)

	def op_gt(self, a, b, c):
		val = 1 if self.lookup(b) > self.lookup(c) else 0
		self.op_set(a, val)

	def op_jmp(self, a):
		self.address = self.lookup(a)
		return 'jmp'

	def op_jt(self, a, b):
		if self.lookup(a) != 0:
			return self.op_jmp(b)

	def op_jf(self, a, b):
		if self.lookup(a) == 0:
			return self.op_jmp(b)

	def op_add(self, a, b, c):
		val = (self.lookup(b) + self.lookup(c)) % (2**15)
		self.op_set(a, val)

	def op_mult(self, a, b, c):
		val = (self.lookup(b) * self.lookup(c)) % (2**15)
		self.op_set(a, val)

	def op_mod(self, a, b, c):
		val = self.lookup(b) % self.lookup(c)
		self.op_set(a, val)

	def op_and(self, a, b, c):
		val = self.lookup(b) & self.lookup(c)
		self.op_set(a, val)

	def op_or(self, a, b, c):
		val = self.lookup(b) | self.lookup(c)
		self.op_set(a, val)

	def op_not(self, a, b):
		val = (-self.lookup(b) - 1) % (2**15)
		self.op_set(a, val)

	def op_rmem(self, a, b):
		val = self.memory[self.lookup(b)]
		self.op_set(a, val)

	def op_wmem(self, a, b):
		val = self.lookup(b)
		self.memory[self.lookup(a)] = val

	def op_call(self, a):
		self.stack.append(self.address + 1)
		return self.op_jmp(a)

	def op_ret(self):
		if len(self.stack) == 0:
			self.op_halt()
		return self.op_jmp(self.stack.pop())

	def op_out(self, a):
		sys.stdout.write(chr(self.lookup(a)))

	def op_in(self, a):
		prompt = "> "

		if self.input_buffer is None:
			if len(self.autoinput) > 0 and self.autoinput[0] != '---':
				command = self.autoinput.popleft()
				print "(auto) %s%s" % (prompt, command)
			else:
				command = raw_input(prompt)
			
			for ac in self.ADMIN_COMMANDS:
				if command.startswith(ac):
					cmd = getattr(self, "admin_%s" % ac.replace(' ', '_'))
					cmd_args = command[len(ac) + 1:].split()
					cmd(*cmd_args)

			self.input_buffer = (c for c in command)

		try:
			char = self.input_buffer.next()
		except StopIteration:
			self.input_buffer = None
			char = "\n"

		self.registers[self.to_register(a)] = ord(char)

	def op_noop(self):
		pass



if __name__ == '__main__':
	mm = MagicMachine()
	mm.load_bin('challenge.bin')
	mm.load_autoinput('autoinput.txt')
	mm.run()

File: magicmachine.rb
require 'pry'

class MagicMachine
	class Halt < StandardError; end

	OPCODES = %w{
		halt set push pop eq gt jmp jt jf add mult mod and or not 
		rmem wmem call ret out in noop
	}

	ADMIN_COMMANDS = [
		'quit', 'registers', 'set register', 'fix teleporter'
	]

	attr_reader :memory, :registers, :stack

	def initialize()
		@address = 0
		@memory = Array.new(2**15){ 0 }
		@registers = Array.new(8){ 0 }
		@stack = []
		@input_buffer = []
		@autoinput = []
	end

	def inspect
		to_s
	end

	def load_bin(filename)
		next_address = 0

		File.open(filename) do |file|
			until file.eof?
				next_vals = file.read(2).bytes.map{ |b| b.ord }
				@memory[next_address] = next_vals[0] + (next_vals[1] * 2**8)
				next_address += 1
			end
		end

		next_address
	end

	def load_autoinput(filename)
		@autoinput = File.open(filename).readlines
	end

	def run(start_address = 0)
		@address = start_address

		while @address < @memory.length
			opaddr = @address
			opnum = @memory[@address]
			opname = OPCODES[opnum]
			operation = method("op_#{opname}")
			opargs = []

			operation.arity.times do
				@address += 1
				opargs << @memory[@address]
			end

			begin
				ret = operation.call(*opargs)
				@address += 1 if ret != 'jmp'
			rescue Halt
				break
			end
		end
	end

	def to_register(value)
		value % (2**15)
	end

	def resolve(value)
		if value >= (2**15)
			@registers[to_register(value)]
		else
			value
		end
	end

	def admin_quit
		op_halt
	end

	def admin_registers
		puts @registers.inspect
	end

	def admin_set_register(regnum, val)
		op_set(regnum.to_i, val.to_i)
	end

	def admin_fix_teleporter
		@registers[7] = 5
		@memory[6027...6030] = [1, 32769, 32775]
		@memory[6030...6034] = [9, 32768, 32769, 1]
	end

	def op_halt
		raise Halt
	end

	def op_set(a, b)
		@registers[to_register(a)] = resolve(b)
	end

	def op_push(a)
		@stack << resolve(a)
	end

	def op_pop(a)
		op_set(a, @stack.pop)
	end

	def op_eq(a, b, c)
		val = resolve(b) == resolve(c) ? 1 : 0
		op_set(a, val)
	end

	def op_gt(a, b, c)
		val = resolve(b) > resolve(c) ? 1 : 0
		op_set(a, val)
	end

	def op_jmp(a)
		@address = resolve(a)
		'jmp'
	end

	def op_jt(a, b)
		op_jmp(b) if resolve(a) != 0
	end

	def op_jf(a, b)
		op_jmp(b) if resolve(a) == 0
	end

	def op_add(a, b, c)
		val = (resolve(b) + resolve(c)) % (2**15)
		op_set(a, val)
	end

	def op_mult(a, b, c)
		val = (resolve(b) * resolve(c)) % (2**15)
		op_set(a, val)
	end

	def op_mod(a, b, c)
		val = resolve(b) % resolve(c)
		op_set(a, val)
	end

	def op_and(a, b, c)
		val = resolve(b) & resolve(c)
		op_set(a, val)
	end

	def op_or(a, b, c)
		val = resolve(b) | resolve(c)
		op_set(a, val)
	end

	def op_not(a, b)
		val = (-resolve(b) - 1) % (2**15)
		op_set(a, val)
	end

	def op_rmem(a, b)
		val = @memory[resolve(b)]
		op_set(a, val)
	end

	def op_wmem(a, b)
		@memory[resolve(a)] = resolve(b)
	end

	def op_call(a)
		@stack << @address + 1
		op_jmp(a)
	end

	def op_ret
		op_halt if @stack.empty?
		op_jmp(@stack.pop)
	end

	def op_out(a)
		print resolve(a).chr
	end

	def op_in(a)
		prompt = "> "

		if @input_buffer.empty?
			command = nil

			while command.nil?
				if @autoinput.empty? || @autoinput[0] == "---\n"
					print prompt
					command = gets
				else
					command = @autoinput.shift

					if command.start_with?('#')
						command = nil
						next
					end

					print "(auto) #{prompt}#{command}"
				end

				ADMIN_COMMANDS.each do |admin_command|
					if command.start_with?(admin_command)
						method_name = "admin_#{admin_command.gsub(' ', '_')}"
						args = command[admin_command.length+1..-1].split
						method(method_name).call(*args)

						command = nil
						break
					end
				end
			end

			@input_buffer = command.split('')
		end

		char = @input_buffer.shift
		op_set(a, char.ord)
	end

	def op_noop
	end

end

if __FILE__ == $0
    mm = MagicMachine.new
    mm.load_bin('challenge.bin')
    mm.load_autoinput('autoinput.txt')
	mm.run
end


